x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java

```diff
@@ -131,9 +136,17 @@ public LifecyclePolicyMetadata read(StreamInput in, String key) {
         );
 
         for (String deletedPolicyName : mapDiff.getDeletes()) {
-            lifecyclePolicyMap.remove(deletedPolicyName);
-            firstStepMap.remove(deletedPolicyName);
-            stepMap.remove(deletedPolicyName);
+            LifecyclePolicyMetadata policyMetadataToRemove = lifecyclePolicyMap.remove(deletedPolicyName);
+            if (policyMetadataToRemove != null) {
+                for (int actionsOrderVersion = policyMetadataToRemove.getPolicy()
+                    .getType()
+                    .getLatestActionsOrderVersion(); actionsOrderVersion >= VERSION_ONE; actionsOrderVersion--) {
+                    // remove all versions for the deleted policy
+                    VersionedPolicyKey key = new VersionedPolicyKey(actionsOrderVersion, policyMetadataToRemove.getName());
+                    firstStepMap.remove(key);
+                    stepMap.remove(key);
+                }
+            }
```

Comment:
I think a comment here might help when we want to understand why we need to loop through all versions.